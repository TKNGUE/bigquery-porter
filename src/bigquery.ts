import type { Metadata } from '@google-cloud/common';
import { BigQuery } from '@google-cloud/bigquery';
import pThrottle from 'p-throttle';
import * as path from 'node:path';

interface BigQueryResource {
  id?: string;
  baseUrl?: string;
  metadata?: Metadata;
  projectId?: string;
  parent?: BigQueryResource;
}

const buildThrottledBigQueryClient = (concurrency: number, interval_limit: number) => {
  const throttle = pThrottle({
    limit: concurrency,
    interval: interval_limit,
  });

  return new Proxy<BigQuery>(
    new BigQuery(),
    {
      get: (obj: BigQuery, sKey: string | symbol) => {
        const member = (obj as any)[sKey];
        // Request Throttling
        if (member instanceof Function && sKey == 'request') {
          return async (...args: any[]) => {
            let result: any;
            await throttle(async () => {
              result = member.apply(obj, args);
            })();
            return result;
          };
        }
        return member;
      },
    },
  );
}

const bq2path = (bqObj: BigQueryResource, asDefaultProject: boolean) => {
  let tree: string[] = [];
  let it: BigQueryResource = bqObj;
  let depth = 0;

  while (true) {
    depth += 1;
    if (it.id) {
      tree.push(it.id);
    }

    if (!it.parent) {
      break;
    }
    it = it.parent;
  }

  if (asDefaultProject) {
    tree.push('@default');
  } else if (bqObj.projectId) {
    tree.push(bqObj.projectId);
  } else if (bqObj.parent?.projectId) {
    tree.push(bqObj.parent.projectId);
  }

  const ns = bqObj.baseUrl?.replace('/', '@');
  if (ns && depth == 3 && !['/tables'].includes(bqObj.baseUrl as string)) {
    tree.splice(1, 0, ns);
  }
  return tree.reverse().join('/');
};

const path2bq = (
  fpath: string,
  rootPath: string,
  defaultProjectId: string
) => {
  const rootDir = path.normalize(rootPath);
  const [catalogId, schemaId, namespace_or_name, name_or_missing] = path
    .dirname(
      path.relative(rootDir, fpath.replace('@default', defaultProjectId)),
    ).split('/');
  const name = name_or_missing ?? namespace_or_name;
  return [catalogId, schemaId, name].filter((n) => n).join('.');
};


const normalizedBQPath = (
  bqPath: string
  , defaultProject?: string
  , isDataset: boolean = false
): string => {
  const cleanedPath = bqPath.replace(/`/g, '');
  const parts = cleanedPath.split('.');

  if (parts.length == 2) {
    if (isDataset) {
      return cleanedPath
    }
    const [dst_schema, dst_name] = parts;
    const dst_project = defaultProject;
    return `${dst_project}.${dst_schema}.${dst_name}`;
  } else if (parts.length == 1) {
    // FIXME: Possible dataset, temporary table or CTE view name
    if (isDataset) {
      const [dst_schema] = parts;
      return `${defaultProject}.${dst_schema}`;
    }
    return cleanedPath
  } else {
    const [dst_project, dst_schema, dst_name] = parts;
    return `${dst_project}.${dst_schema}.${dst_name}`;
  }
};

export {
  BigQueryResource,
  bq2path,
  path2bq,
  normalizedBQPath,
  buildThrottledBigQueryClient
};
