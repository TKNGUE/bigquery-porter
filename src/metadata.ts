import * as path from 'node:path';
import * as fs from 'node:fs';

import type { ServiceObject } from '@google-cloud/common';
import { Dataset, Model, Routine, Table } from '@google-cloud/bigquery';

import { fetchRowAccessPolicy } from '../src/rowAccessPolicy.js';

const jsonSerializer = (obj: any) => JSON.stringify(obj, null, 2);

// type Labels = Map<string, string>;
const syncMetadata = async (
  bqObject: Dataset | Table | Routine | Model,
  dirPath: string,
  options?: { versionhash?: string; push?: boolean },
): Promise<string[]> => {
  type systemDefinedLabels = {
    'bqport-versionhash': string;
  };

  const metadataPath = path.join(dirPath, 'metadata.json');
  const readmePath = path.join(dirPath, 'README.md');
  const fieldsPath = path.join(dirPath, 'schema.json');
  const syncLabels: systemDefinedLabels = {
    'bqport-versionhash': `${Math.floor(Date.now() / 1000)}-${
      options?.versionhash ?? 'HEAD'
    }`,
  };
  const jobs: Promise<any>[] = [];

  const projectId = bqObject.metadata?.datasetReference?.projectId ??
    (bqObject.parent as ServiceObject).metadata?.datasetReference?.projectId;

  const apiQuery: { projectId: string; location?: string } = { projectId };
  if (bqObject instanceof Dataset && bqObject.location) {
    apiQuery.location = bqObject.location;
  }
  // const location =  ?  : (bqObject.parent as Dataset).location;
  const [metadata] = await bqObject.getMetadata(apiQuery);

  const newMetadata = Object.fromEntries(
    Object.entries({
      type: metadata.type,
      // etag: metadata.etag,
      routineType: metadata.routineType,
      modelType: metadata.modelType,

      // Filter predefined labels
      labels: Object.fromEntries(
        Object.entries(metadata?.labels ?? [])
          .filter(([k]) => !(k in syncLabels)),
      ),
      // Table attribute
      timePartitioning: metadata?.timePartitioning,
      clustering: metadata?.clustering,

      // Snapshot attribute
      snapshotDefinition: metadata?.snapshotDefinition,

      // cloneDefinition attribute
      cloneDefinition: metadata?.snapshotDefinition,

      // Dataset attribute
      access: metadata?.access,
      location: bqObject instanceof Dataset ? metadata?.location : undefined,

      // Routine Common atributes
      language: metadata?.language,
      arguments: metadata?.arguments,

      // SCALAR FUNCTION / PROCEDURE attribute
      returnType: metadata?.returnType,

      // TABLE FUNCTION attribute
      returnTableType: metadata?.returnTableType,

      // MODEL attribute
      featureColumns: metadata?.featureColumns,
      labelColumns: metadata?.labelColumns,
      trainingRuns: metadata?.trainingRuns,
    })
      // Remove invalid fields
      .filter(([_, v]) => !!v && Object.keys(v).length > 0),
  );

  /* -----------------------------
  / Merge local and remote metadata
  */
  if (fs.existsSync(metadataPath)) {
    const local = await fs.promises.readFile(metadataPath)
      .then((s) => JSON.parse(s.toString()));

    if (options?.push) {
      // TODO: Add allowlist check to write
      Object.entries(newMetadata).forEach(([attr]) => {
        newMetadata[attr] = local[attr];
      });
    } else {
      Object.entries(newMetadata).forEach(([attr]) => {
        newMetadata[attr] = local[attr] ?? metadata[attr];
      });
    }
  }

  // schema.json: local file <---> BigQuery Table
  if (metadata?.schema?.fields) {
    // Merge upstream and downstream schema description
    // due to some bigquery operations like view or materialized view purge description
    if (fs.existsSync(fieldsPath)) {
      const oldFields = await fs.promises.readFile(fieldsPath)
        .then((s) => JSON.parse(s.toString()))
        .catch((err: Error) => console.error(err));
      // Update
      Object.entries(metadata.schema.fields).map(
        ([k, v]: [string, any]) => {
          if (
            k in oldFields &&
            metadata.schema.fields[k].description &&
            metadata.schema.fields[k].description != v.description
          ) {
            metadata.schema.fields[k].description = v.description;
          }
        },
      );
    }

    jobs.push(fs.promises.writeFile(
      fieldsPath,
      jsonSerializer(metadata.schema.fields),
    ));
  }

  // README.md
  if (metadata.description !== undefined) {
    const upstream = metadata.description;
    newMetadata['description'] = upstream;
    if (fs.existsSync(readmePath)) {
      const local = await fs.promises.readFile(readmePath)
        .then((s) => s.toString());
      newMetadata['description'] = local;
      if (upstream !== undefined && local != upstream) {
        console.warn(
          'Warning: Local README.md file cannot be updated due to already exists. Please remove README.md file to update it.',
        );
      }
    }

    if (newMetadata['description'] !== undefined) {
      jobs.push(fs.promises.writeFile(readmePath, newMetadata['description']));
    }
  }

  if (options?.push) {
    jobs.push((bqObject as any).setMetadata(newMetadata));
  }

  // metadata.json
  const rowAccessPolicies = bqObject instanceof Table
    ? (await fetchRowAccessPolicy(
      bqObject.bigQuery,
      metadata?.tableReference.datasetId,
      metadata?.tableReference.tableId,
    ).catch((err: Error) => {
      console.warn(err.message);
      return [];
    }))
    : [];

  if (rowAccessPolicies.length > 0) {
    const entryRowAccessPolicies = rowAccessPolicies.map((p) => ({
      policyId: p.rowAccessPolicyReference.policyId,
      filterPredicate: p.filterPredicate,
    }));
    jobs.push(
      fs.promises.writeFile(
        metadataPath,
        jsonSerializer({
          ...newMetadata,
          rowAccessPolicies: entryRowAccessPolicies,
        }),
      ),
    );
  } else {
    jobs.push(
      fs.promises.writeFile(
        metadataPath,
        jsonSerializer(newMetadata),
      ),
    );
  }

  await Promise.all(jobs);
  return [];
};

export { syncMetadata };
