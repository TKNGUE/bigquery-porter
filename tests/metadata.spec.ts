import {
  afterAll,
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  expect,
  it,
} from 'vitest';
import { tmpdir } from 'node:os';
import * as fs from 'node:fs';
import * as path from 'node:path';

import { BigQuery } from '@google-cloud/bigquery';
import { syncMetadata } from '../src/metadata.js';

describe('syncMetadata: Pull', () => {
  const bqClient = new BigQuery();
  const _dataset = fs.mkdtempSync(`${tmpdir()}${path.sep}`);
  const _resource = 'resource';
  const dPath = path.join(_dataset, _resource);

  afterAll(async () => {
    fs.rmdirSync(_dataset, { recursive: true });
  });

  beforeEach(async () => {
    fs.mkdirSync(path.join(_dataset, _resource), { recursive: true });
  });

  afterEach(async () => {
    fs.rmdirSync(path.join(_dataset, _resource), { recursive: true });
  });

  it('for Table: bigquery-public-data:austin_bikeshare.bikeshare_stations', async () => {
    const table = bqClient.dataset('austin_bikeshare', {
      projectId: 'bigquery-public-data',
    }).table(
      'bikeshare_stations',
    );

    // test
    const modified = (await syncMetadata(table, dPath))
      .map((n) => path.basename(n));
    expect(modified.length).toBe(3);

    const _load = (f: string) => fs.readFileSync(path.join(dPath, f), 'utf-8');
    expect(modified).toContain('metadata.json');
    expect(fs.existsSync(path.join(dPath, 'metadata.json'))).toBe(true);
    expect(JSON.parse(_load('metadata.json')))
      .toMatchSnapshot();

    expect(modified).toContain('schema.json');
    expect(fs.existsSync(path.join(dPath, 'schema.json'))).toBe(true);
    expect(JSON.parse(_load('schema.json')))
      .toMatchSnapshot();

    expect(modified).toContain('README.md');
    expect(fs.existsSync(path.join(dPath, 'README.md'))).toBe(true);
    expect(_load('README.md'))
      .toMatchSnapshot();
  });

  it('for Routine: bqutil.fn.int', async () => {
    const table = bqClient.dataset('fn', { projectId: 'bqutil' }).routine(
      'int',
    );

    const modified = (await syncMetadata(table, dPath))
      .map((n) => path.basename(n));

    const _load = (f: string) => fs.readFileSync(path.join(dPath, f), 'utf-8');

    expect(modified.length).toBe(1);

    expect(modified).toContain('metadata.json');
    expect(fs.existsSync(path.join(dPath, 'metadata.json'))).toBe(true);
    expect(JSON.parse(_load('metadata.json')))
      .toMatchSnapshot();

    expect(fs.existsSync(path.join(dPath, 'schema.json'))).toBe(false);

    expect(fs.existsSync(path.join(dPath, 'README.md'))).toBe(false);
  });
});

describe('syncMetadata: Push', () => {
  const bqClient = new BigQuery();
  const _local = fs.mkdtempSync(`${tmpdir()}${path.sep}`);

  const _dataset = 'zz' + path.basename(_local);
  const _fsResource = _local;

  // Utils
  const toPath = (p: string) => path.join(_fsResource, p);
  const _load = (f: string) =>
    fs.readFileSync(path.join(_fsResource, f), 'utf-8');
  const _write = (p: string, content: string) =>
    fs.promises.writeFile(
      path.join(_fsResource, p),
      content,
      'utf-8',
    );

  beforeAll(async () => {
    const [dataset] = await bqClient.createDataset(_dataset, {});
    // https://cloud.google.com/bigquery/docs/reference/v2/tables#resource
    await dataset.createTable('sample_table', {
      schema: 'id:integer, name:string',
      description: 'sample table',
    });
  });

  afterAll(async () => {
    await bqClient.dataset(_dataset).delete({ force: true });
    fs.rmdirSync(_local, { recursive: true });
  });

  it('for Table: sample_table', async () => {
    const table = bqClient.dataset(_dataset).table(
      'sample_table',
    );
    const afterDescription = 'Updated description';
    const afterSchema = [
      {
        name: 'id',
        type: 'INTEGER',
        description: 'Updated description for \'id\'',
      },
      {
        name: 'name',
        type: 'STRING',
      },
    ];

    // test
    await Promise.all([
      _write('README.md', afterDescription),
      _write('schema.json', JSON.stringify(afterSchema)),
    ]);

    const modified = (await syncMetadata(table, _fsResource, { push: true }))
      .map((n) => path.basename(n));

    expect(modified.length).toBe(3);
    const [metadata] = await table.getMetadata();

    expect(fs.existsSync(toPath('README.md'))).toBe(true);
    expect(_load('README.md'))
      .toBe(afterDescription);

    expect(fs.existsSync(toPath('schema.json'))).toBe(true);
    expect(JSON.parse(_load('schema.json')))
      .toMatchSnapshot();
    expect(metadata.schema.fields).toMatchObject(afterSchema);

    expect(fs.existsSync(path.join(toPath('metadata.json')))).toBe(true);
    expect(JSON.parse(_load('metadata.json')))
      .toMatchSnapshot();
  });
});
