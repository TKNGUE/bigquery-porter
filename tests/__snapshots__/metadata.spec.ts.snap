// Vitest Snapshot v1

exports[`syncMetadata: Pull > for Routine: bqutil.fn.int 1`] = `
{
  "arguments": [
    {
      "argumentKind": "ANY_TYPE",
      "name": "v",
    },
  ],
  "language": "SQL",
  "routineType": "SCALAR_FUNCTION",
}
`;

exports[`syncMetadata: Pull > for Table: bigquery-public-data:austin_bikeshare.bikeshare_stations 1`] = `
{
  "type": "TABLE",
}
`;

exports[`syncMetadata: Pull > for Table: bigquery-public-data:austin_bikeshare.bikeshare_stations 2`] = `
[
  {
    "mode": "NULLABLE",
    "name": "station_id",
    "type": "INTEGER",
  },
  {
    "mode": "NULLABLE",
    "name": "name",
    "type": "STRING",
  },
  {
    "mode": "NULLABLE",
    "name": "status",
    "type": "STRING",
  },
  {
    "mode": "NULLABLE",
    "name": "address",
    "type": "STRING",
  },
  {
    "mode": "NULLABLE",
    "name": "alternate_name",
    "type": "STRING",
  },
  {
    "mode": "NULLABLE",
    "name": "city_asset_number",
    "type": "INTEGER",
  },
  {
    "mode": "NULLABLE",
    "name": "property_type",
    "type": "STRING",
  },
  {
    "mode": "NULLABLE",
    "name": "number_of_docks",
    "type": "INTEGER",
  },
  {
    "mode": "NULLABLE",
    "name": "power_type",
    "type": "STRING",
  },
  {
    "mode": "NULLABLE",
    "name": "footprint_length",
    "type": "INTEGER",
  },
  {
    "mode": "NULLABLE",
    "name": "footprint_width",
    "type": "FLOAT",
  },
  {
    "mode": "NULLABLE",
    "name": "notes",
    "type": "STRING",
  },
  {
    "mode": "NULLABLE",
    "name": "council_district",
    "type": "INTEGER",
  },
  {
    "mode": "NULLABLE",
    "name": "modified_date",
    "type": "TIMESTAMP",
  },
]
`;

exports[`syncMetadata: Pull > for Table: bigquery-public-data:austin_bikeshare.bikeshare_stations 3`] = `"Austin Bikeshare Stations table"`;

exports[`syncMetadata: Push > for Table: sample_table 1`] = `
[
  {
    "description": "Updated description for 'id'",
    "name": "id",
    "type": "INTEGER",
  },
  {
    "name": "name",
    "type": "STRING",
  },
]
`;

exports[`syncMetadata: Push > for Table: sample_table 2`] = `
{
  "type": "TABLE",
}
`;
